WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+			ImmutableArray<RequestNode> negative = Enumerable.Reverse(ordered.ThenByDescending(x => InDegree(x, credentialType)).Where(v => Balance(v, credentialType) < 0)).ToImmutableArray();
+
+			if (negative.Length == 0)
+			{
+				largestMagnitudeNode = null;
+				smallMagnitudeNodes = null;
+				return false;
+			}
+
+			var nPositive = 1;
+			var nNegative = 1;
+
+			IEnumerable<RequestNode> posCandidates() => positive.Take(nPositive);
+			IEnumerable<RequestNode> negCandidates() => negative.Take(nNegative);
+			long posSum() => posCandidates().Sum(x => Balance(x, credentialType));
+			long negSum() => negCandidates().Sum(x => Balance(x, credentialType));
+			long compare() => posSum().CompareTo(-1 * negSum());
+
+			// We want to fully discharge the larger (in absolute magnitude) of
+			// the two nodes, so we will add more nodes to the smaller one until
+			// we can fully cover. At each step of the iteration we fully
+			// discharge at least 2 nodes from the queue.
+			var initialComparison = compare();
+			var fanIn = initialComparison == -1;
+
+			if (initialComparison != 0)
+			{
+				Action takeOneMore = fanIn ? () => nPositive++ : () => nNegative++;
+
+				// Take more nodes until the comparison sign changes or
+				// we run out.
+				while (initialComparison == compare()
+				       && (fanIn ? positive.Length - nPositive > 0
+				                 : negative.Length - nNegative > 0))
+				{
+					takeOneMore();
+				}
+			}
+
+			largestMagnitudeNode = (fanIn ? negative.First() : positive.First() );
+			smallMagnitudeNodes = (fanIn ? positive.Take(nPositive).Reverse() : negative.Take(nNegative)); // reverse positive values so we always proceed in order of increasing magnitude
+
+			return true;
+		}
+
+		// Build a k-ary tree bottom up to reduce a list of nodes.
+		private ImmutableList<RequestNode> LimitCount(IEnumerable<RequestNode> nodes, int maxCount, CredentialType credentialType)
+		{
+			var nodeQueue = nodes.ToImmutableList();
+
+			while (nodeQueue.Count > maxCount)
+			{
+				// Replace up to k nodes, possibly the entire queue, with a
+				// single reissuance node which combines their values. The total
+				// number of items might be less than K but still larger than
+				// maxCount.
+				var take = Math.Min(K, nodeQueue.Count);
+				var reissuance = NewReissuanceNode();
+				DrainReissuance(reissuance, nodeQueue.Take(take), credentialType);
+				nodeQueue = nodeQueue.RemoveRange(0, take).Add(reissuance);
+			}
+
+			return nodeQueue;
+		}
+
+		private RequestNode NewReissuanceNode()
+		{
+			var node = new RequestNode(Vertices.Count, Enumerable.Repeat(0L, K).ToImmutableArray());
+			AddNode(node);
+			return node;
+		}
+
+		// Drain values into a reissuance request (towards the center of the graph).
+		private void DrainReissuance(RequestNode x, IEnumerable<RequestNode> ys, CredentialType type)
```

Comment:
`xs`, `ys` are not part of the C# devs jardon.

Reply:
hmm, after I split this into two methods i think `reissuanceNode` makes sense instead of `x`, and maybe `nodes` or `terminalNodes` instead of `ys`